### ISSUE TITLE
Redundant closure gives invalid suggestion

### ISSUE THREAD

**Issue by rcoh** [rust-lang/rust-clippy#7218]
Lint name: `redundant_closure`

I tried this code:

```rust
use std::env::VarError;
fn main() {
    
    let _x: Box<dyn Fn(&str) -> Result<String, VarError> + Send + Sync> = 
        Box::new(|key|std::env::var(key));
        
        
    // does not compile:
    // let _x: Box<dyn Fn(&str) -> Result<String, VarError> + Send + Sync> = 
    //    Box::new(std::env::var);
}
```

I expected to see this happen: Clippy should not suggest `redundant_closure`, because of lifetime inference, the closure is required:
```
   Compiling playground v0.0.1 (/playground)
error: implementation of `FnOnce` is not general enough
  --> src/main.rs:10:9
   |
10 |         Box::new(std::env::var);
   |         ^^^^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: `fn(&'2 str) -> Result<String, VarError> {var::<&'2 str>}` must implement `FnOnce<(&'1 str,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 str,)>`, for some specific lifetime `'2`
   ```

### Meta

- `cargo clippy -V`: `clippy 0.1.52 (9bc8c42b 2021-05-09) (also 0.1.54 on playground)`
- `rustc -Vv`: 

```
rustc 1.52.1 (9bc8c42bb 2021-05-09)
binary: rustc
commit-hash: 9bc8c42bb2f19e745a63f3445f1ac248fb015e53
commit-date: 2021-05-09
host: x86_64-apple-darwin
release: 1.52.1
LLVM version: 12.0.0
```

**Comment by Jarcho** [comment]
This was fixed by #7661.
